File: simulator/simulator.py
import os
import sys
import argparse
import time
import json
import urllib.request
import urllib.error

# Fix Windows console UTF-8 UnicodeEncodeError
if sys.platform == "win32":
    try:
        sys.stdout.reconfigure(encoding='utf-8', errors='replace')
        sys.stderr.reconfigure(encoding='utf-8', errors='replace')
    except Exception:
        pass

# Ensure simulator directory is in sys.path regardless of working directory
_SIMULATOR_DIR = os.path.dirname(os.path.abspath(__file__))
if _SIMULATOR_DIR not in sys.path:
    sys.path.insert(0, _SIMULATOR_DIR)

try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False

from config import STATIONS, TELEMETRY_ENDPOINT, ALERT_ENDPOINT, TELEMETRY_INTERVAL_SECONDS
from sensors import StationSensorModel
from scenarios import run_generator_thermal_runaway, run_blizzard_warning

def _http_post(url, payload):
    """Resilient HTTP POST supporting both requests and urllib standard library."""
    if HAS_REQUESTS:
        try:
            res = requests.post(url, json=payload, timeout=2.0)
            return res.status_code, res.text
        except Exception:
            pass

    # Fallback to urllib.request
    try:
        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            url,
            data=data,
            headers={'Content-Type': 'application/json'},
            method='POST'
        )
        with urllib.request.urlopen(req, timeout=2.0) as res:
            return res.status, res.read().decode('utf-8')
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode('utf-8')
    except Exception as e:
        return 0, str(e)

def send_telemetry(payload):
    st_name = "MAITRI" if "maitri" in payload.get('station_id', '') else "BHARATI"
    temp = payload.get('temperature', 0)
    gen_temp = payload.get('generator_temperature', 0)
    batt = payload.get('battery_level') or payload.get('battery') or 0

    endpoints_to_try = [
        TELEMETRY_ENDPOINT,
        "http://localhost:5000/api/telemetry",
        "http://localhost:5000/api/sensor-data",
        "http://localhost:8000/api/telemetry",
    ]
    # Remove duplicate endpoints preserving order
    seen = set()
    endpoints = [x for x in endpoints_to_try if not (x in seen or seen.add(x))]

    for endpoint in endpoints:
        status_code, text = _http_post(endpoint, payload)
        if status_code in [200, 201, 202]:
            print(f"[SIMULATOR] Station: {st_name} | Temp: {temp}°C | Gen: {gen_temp}°C | Battery: {batt}% | Status: {status_code} OK (-> {endpoint})", flush=True)
            return
    
    print(f"[SIMULATOR ERROR] Connection failed for {st_name} to endpoints: {endpoints}", flush=True)

def send_alert(payload):
    st_name = "MAITRI" if "maitri" in payload.get('station_id', '') else "BHARATI"
    endpoints_to_try = [
        ALERT_ENDPOINT,
        "http://localhost:5000/api/alerts",
        "http://localhost:8000/api/alerts",
    ]
    seen = set()
    endpoints = [x for x in endpoints_to_try if not (x in seen or seen.add(x))]

    for endpoint in endpoints:
        status_code, text = _http_post(endpoint, payload)
        if status_code in [200, 201, 202]:
            print(f"[SIMULATOR ALERT] 🚨 {st_name} PRIORITY ALERT DISPATCHED: {payload.get('priority')} -> {payload.get('message')}", flush=True)
            return

    print(f"[SIMULATOR ERROR] Alert dispatch failed to endpoints: {endpoints}", flush=True)

def main():
    parser = argparse.ArgumentParser(description="Antarctic Research Station Sensor Simulator")
    parser.add_argument("--scenario", choices=["generator_overheat", "blizzard", "none"], default="none",
                        help="Trigger a specific emergency scenario immediately")
    parser.add_argument("--station", choices=["station-maitri", "station-bharati"], default="station-maitri",
                        help="Target station for scenario")
    parser.add_argument("--interval", type=float, default=TELEMETRY_INTERVAL_SECONDS,
                        help="Telemetry interval in seconds")
    args = parser.parse_args()

    # Initialize sensor models for all stations
    station_models = {
        s_id: StationSensorModel(s_id, s_meta)
        for s_id, s_meta in STATIONS.items()
    }

    print("==================================================================", flush=True)
    print("🏔️  [SIMULATOR] Starting POLARIS Station Telemetry Simulator", flush=True)
    print(f"📡 Target Endpoint: {TELEMETRY_ENDPOINT}", flush=True)
    print(f"🚨 Alert Endpoint:  {ALERT_ENDPOINT}", flush=True)
    print("==================================================================", flush=True)

    # If a specific scenario is requested
    if args.scenario == "generator_overheat":
        target_model = station_models[args.station]
        run_generator_thermal_runaway(target_model, send_telemetry, send_alert)
        return
    elif args.scenario == "blizzard":
        target_model = station_models[args.station]
        run_blizzard_warning(target_model, send_telemetry, send_alert)
        return

    # Continuous background telemetry loop
    print("\n▶️  [SIMULATOR] Starting normal multi-station telemetry loop...\n", flush=True)
    try:
        step_count = 0
        while True:
            step_count += 1
            for station_id, model in station_models.items():
                telemetry = model.step_normal_drift()
                send_telemetry(telemetry)

            time.sleep(args.interval)

    except KeyboardInterrupt:
        print("\n⏹️ [SIMULATOR] Stopped by operator.", flush=True)
        sys.exit(0)

if __name__ == "__main__":
    main()

File: simulator/simulator.js
/**
 * POLARIS Antarctic Station Sensor Simulator.
 * Provides continuous telemetry drift & emergency scenario simulations for Maitri and Bharati.
 */

const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:5000';
const TELEMETRY_ENDPOINT = `${BACKEND_URL}/api/telemetry`;
const ALERT_ENDPOINT = `${BACKEND_URL}/api/alerts`;

const STATIONS = {
  'station-maitri': {
    name: 'Maitri Research Station',
    state: {
      temperature: -18.7,
      battery: 74.0,
      battery_level: 74.0,
      power_consumption: 105.0,
      generator_status: 'RUNNING',
      generator_temperature: 78.4,
      wind_speed: 28.0,
      water_level: 88.0,
      comms_status: 'SAT_LINK_NOMINAL',
    }
  },
  'station-bharati': {
    name: 'Bharati Research Station',
    state: {
      temperature: -14.2,
      battery: 91.0,
      battery_level: 91.0,
      power_consumption: 148.0,
      generator_status: 'RUNNING',
      generator_temperature: 74.1,
      wind_speed: 44.0,
      water_level: 94.0,
      comms_status: 'SAT_LINK_NOMINAL',
    }
  }
};

async function postJson(url, data) {
  const targets = [url];
  if (url.includes('5000')) targets.push(url.replace('5000', '8000'));
  else if (url.includes('8000')) targets.push(url.replace('8000', '5000'));

  if (url.includes('/api/telemetry')) targets.push(url.replace('/api/telemetry', '/api/sensor-data'));

  for (const target of targets) {
    try {
      const res = await fetch(target, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data),
      });
      if (res.ok) {
        return await res.json();
      }
    } catch (err) {
      // try next target
    }
  }
  console.error(`[SIMULATOR ERROR] Connection refused for: ${url}`);
  return null;
}

async function runGeneratorThermalRunaway(stationId = 'station-maitri') {
  const stn = STATIONS[stationId];
  const stName = stationId.includes('maitri') ? 'MAITRI' : 'BHARATI';
  console.log(`\n🔥 [SIMULATOR] Initiating Generator Thermal Runaway Scenario for ${stName}...`);
  const steps = [70.0, 78.0, 85.0, 92.0, 95.0];

  for (let i = 0; i < steps.length; i++) {
    const temp = steps[i];
    stn.state.generator_temperature = temp;
    stn.state.generator_status = temp >= 95 ? 'CRITICAL' : temp >= 85 ? 'WARNING' : 'RUNNING';

    const payload = {
      station_id: stationId,
      station_name: stn.name,
      ...stn.state,
      recorded_at: new Date().toISOString(),
    };

    console.log(`[SIMULATOR] Station: ${stName} | Step ${i + 1}/5 | Sending telemetry... | Temp: ${payload.temperature}°C | Gen: ${temp}°C | Status: ${stn.state.generator_status}`);
    await postJson(TELEMETRY_ENDPOINT, payload);

    if (temp >= 95.0) {
      console.log(`[SIMULATOR ALERT] 🚨 ${stName} CRITICAL ALERT: Generator failure risk at 95°C!`);
      const alertPayload = {
        station_id: stationId,
        priority: 'CRITICAL',
        category: 'GENERATOR',
        message: `🚨 ${stName} GENERATOR FAILURE RISK: Core temperature breached 95°C! Immediate shutdown required.`,
        sensor_key: 'generator_temperature',
        sensor_value: temp,
        threshold_value: 90.0,
        action_required: 'Switch to Generator 2 and initiate non-critical load shedding.',
        triggered_at: new Date().toISOString(),
      };
      await postJson(ALERT_ENDPOINT, alertPayload);
    }

    await new Promise((r) => setTimeout(r, 1400));
  }

  console.log(`✅ [SIMULATOR] Generator Thermal Runaway completed.\n`);
}

function driftSensors(stationId) {
  const stn = STATIONS[stationId];
  const s = stn.state;

  s.temperature = +(s.temperature + (Math.random() * 0.4 - 0.2)).toFixed(2);
  s.battery_level = +(Math.max(60, Math.min(100, s.battery_level + (Math.random() * 0.2 - 0.1)))).toFixed(2);
  s.power_consumption = +(Math.max(50, Math.min(180, s.power_consumption + (Math.random() * 1.2 - 0.6)))).toFixed(2);
  s.generator_temperature = +(Math.max(68, Math.min(82, s.generator_temperature + (Math.random() * 0.6 - 0.3)))).toFixed(2);
  s.wind_speed = +(Math.max(10, Math.min(85, s.wind_speed + (Math.random() * 2.0 - 1.0)))).toFixed(2);
  s.water_level = +(Math.max(50, Math.min(100, s.water_level + (Math.random() * 0.1 - 0.05)))).toFixed(2);

  return {
    station_id: stationId,
    station_name: stn.name,
    ...s,
    recorded_at: new Date().toISOString(),
  };
}

async function main() {
  const args = process.argv.slice(2);
  const isScenario = args.includes('--scenario') || args.includes('-s') || args.includes('generator_overheat');

  console.log('==================================================================');
  console.log('🏔️  [SIMULATOR] Starting POLARIS Station Telemetry Simulator');
  console.log(`📡 Telemetry Target: ${TELEMETRY_ENDPOINT}`);
  console.log(`🚨 Alert Target:     ${ALERT_ENDPOINT}`);
  console.log('==================================================================');

  if (isScenario) {
    await runGeneratorThermalRunaway('station-maitri');
    return;
  }

  console.log('▶️  [SIMULATOR] Starting normal multi-station telemetry stream (Press Ctrl+C to stop)...\n');
  
  const tick = async () => {
    for (const stnId of Object.keys(STATIONS)) {
      const payload = driftSensors(stnId);
      const stName = stnId.includes('maitri') ? 'MAITRI' : 'BHARATI';
      const res = await postJson(TELEMETRY_ENDPOINT, payload);
      if (res) {
        console.log(`[SIMULATOR] Station: ${stName} | Sending telemetry... | Temp: ${payload.temperature}°C | Gen: ${payload.generator_temperature}°C | Battery: ${payload.battery_level}% | Response: 200 OK`);
      }
    }
  };

  await tick();
  setInterval(tick, 3000);
}

main();
